### PR TITLE
shortcut.xml file saved on modification with comments

### PR DESCRIPTION
**Pull request by cmeriaux**
shortcut.xml file isn't saved on exit but on shortcut modification or macro modification.

Now, it's possible to edit in notepad++ the file shortcut.xml then restart notepad++ (same procedure as contextMenu.xml file) 

shortcut.xml file also contains a description of each shortcut . Example : 

```xml
<NotepadPlus>
    <InternalCommands>
        <!--Appliquer le style 1 - Ctrl+1-->
        <Shortcut id="43022" Ctrl="yes" Alt="no" Shift="no" Key="49" />
        <!--Appliquer le style 2 - Ctrl+2-->
        <Shortcut id="43024" Ctrl="yes" Alt="no" Shift="no" Key="50" />
        <!--Appliquer le style 3 - Ctrl+3-->
        <Shortcut id="43026" Ctrl="yes" Alt="no" Shift="no" Key="51" />
        <!--Nouveau - Ctrl+N-->
        <Shortcut id="41001" Ctrl="yes" Alt="no" Shift="no" Key="78" />
    </InternalCommands>
    <Macros>
        <!--Trim Trailing and save - Ctrl+Shift+S -->
        <Macro name="Trim Trailing and save" Ctrl="yes" Alt="no" Shift="yes" Key="83">
            <Action type="2" message="0" wParam="42024" lParam="0" sParam="" />
            <Action type="2" message="0" wParam="41006" lParam="0" sParam="" />
        </Macro>
    </Macros>
       <UserDefinedCommands>
        <!--Launch in Firefox - Ctrl+Alt+Shift+X -->
        <Command name="Launch in Firefox" Ctrl="yes" Alt="yes" Shift="yes" Key="88">firefox &quot;$(FULL_CURRENT_PATH)&quot;</Command>
        <!--Launch in IE - Ctrl+Alt+Shift+I -->
        <Command name="Launch in IE" Ctrl="yes" Alt="yes" Shift="yes" Key="73">iexplore &quot;$(FULL_CURRENT_PATH)&quot;</Command>
    </UserDefinedCommands>
    <PluginCommands>
        <!--moduleName=CodeAlignmentNpp.dll command=Align by...-->
        <PluginCommand moduleName="CodeAlignmentNpp.dll" internalID="0" Ctrl="yes" Alt="yes" Shift="yes" Key="187" />
        <!--moduleName=DSpellCheck.dll command=Spell Check Document Automatically-->
        <PluginCommand moduleName="DSpellCheck.dll" internalID="0" Ctrl="no" Alt="no" Shift="no" Key="0" />
    </PluginCommands>
    <ScintillaKeys>
        <!--SCI_CUT - Ctrl+X-->
        <ScintKey ScintID="2177" menuCmdID="42001" Ctrl="yes" Alt="no" Shift="no" Key="88">
            <!--SCI_CUT - Ctrl+X-->
            <NextKey Ctrl="no" Alt="no" Shift="yes" Key="46" />
        </ScintKey>
    </ScintillaKeys>
```

It's also fixing this bug : 

Consider that you have some plugins with marco shortcut configured. Then :

- Start Notepad++ with the option -noPlugin
- Stop Notepad++. 
During the close, it rewrites the shortcut.xml. As there is no plugin loaded, the plugin section of shortcut.xml is empty.
- start Notepad++ again. **All plugin shortcut have been loosed**.

### CODE REVIEW

File: PowerEditor/src/Parameters.cpp [hunk 1]
```diff
@@ -2792,6 +2792,16 @@ void NppParameters::writeUserDefinedLang()
 
 void NppParameters::insertCmd(TiXmlNode *shortcutsRoot, const CommandShortcut & cmd)
 {
+	//insert a comment about which command is written in the XML file
+	TiXmlNode *sc2 = shortcutsRoot->InsertEndChild(TiXmlComment());
+	generic_string str = TEXT("");
```
Comment (ariccio) [inline PowerEditor/src/Parameters.cpp]: Why the explicit ` = TEXT("")`? `std::wstring` already default constructs.

Explicitly initializing it adds (a *really* tiny amount of) code bloat.

File: PowerEditor/src/Parameters.cpp [hunk 2]
```diff
@@ -2848,6 +2891,19 @@ void NppParameters::insertPluginCmd(TiXmlNode *pluginCmdRoot, const PluginCmdSho
 
 void NppParameters::insertScintKey(TiXmlNode *scintKeyRoot, const ScintillaKeyMap & scintKeyMap)
 {
+	TiXmlNode * keyComment;
+	keyComment = scintKeyRoot->InsertEndChild(TiXmlComment());
```
Comment (ariccio) [inline PowerEditor/src/Parameters.cpp]: Why the `C89` declare-then-assign style?

`TiXmlNode* keyComment = scintKeyRoot->InsertEndChild(TiXmlComment());` might be clearer?

Or, better yet:

`TiXmlNode* const keyComment = scintKeyRoot->InsertEndChild(TiXmlComment());` :)

File: PowerEditor/src/Notepad_plus.cpp [hunk 3]
```diff
@@ -3947,7 +3947,8 @@ bool Notepad_plus::saveScintillaParams()
 
 bool Notepad_plus::addCurrentMacro()
 {
-	vector<MacroShortcut> & theMacros = (NppParameters::getInstance())->getMacroList();
+	NppParameters * pNppParam = NppParameters::getInstance();
+	vector<MacroShortcut> & theMacros = pNppParam->getMacroList();
 
 	int nbMacro = theMacros.size();
```
Comment (ariccio) [inline PowerEditor/src/Notepad_plus.cpp]: addCurrentMacro has three levels of indentation. Is a drive-by fix in order?

i.e. Changing 

```
if (ms.doDialog() != -1)
{
```
<huge block of code>
```
}
return false;
```

To:

```
if (ms.doDialog() == -1)
{
    return false;
}
```

Comment (ariccio) [inline PowerEditor/src/Notepad_plus.cpp]: Side note, ["drive by fix"](http://www.google.com/search?q=code+%22drive+by+fix%22) is a bit of [microspeak](http://blogs.msdn.com/b/oldnewthing/archive/tags/microspeak/).

File: PowerEditor/src/Parameters.cpp (hunk 1 above)
Comment (cmeriaux) [inline PowerEditor/src/Parameters.cpp]: Ok. I'm a C++ beginner. Thank you for the advise

File: PowerEditor/src/Parameters.cpp (hunk 2 above)
Comment (cmeriaux) [inline PowerEditor/src/Parameters.cpp]: done

File: PowerEditor/src/Notepad_plus.cpp (hunk 3 above)
Comment (cmeriaux) [inline PowerEditor/src/Notepad_plus.cpp]: Sorry, I don't want to make to much modification or Don Ho will not accept my PR.

Comment (ariccio) [inline PowerEditor/src/Notepad_plus.cpp]: Ok. Fair enough.

**Comment by cmeriaux** [comment]
@milipili, hello. Do you find to time to have a look to my PR ? I pushed a new version to rebase with latest commit and I added a menu ""Edit shortcuts config file".
If you can have a look. 
Thanks.

**Comment by milipili** [comment]
I will do it as soon as I can, probably within the week (sorry I was a bit busy somewhere else)

**Comment by cmeriaux** [comment]
@milipili @donho , hello, I'm back from holidays. and  I pushed a new version to rebase with latest commit.
Can You please have a look to my PR ? I can simplify it if necessary. 

It's been 4 month that I'm waiting for this PR to be merged. Please, Tell me what to do if necessary.
Cheers

**Comment by sanastasiou** [comment]
I'm also interested in this one! :) Will allow my plugin to implement shortcuts!

**Comment by donho** [comment]
I'm not for merging the PR because it doesn't fix the bug but get around the bug by changing the current behaviour. I prefer fix the bug and keep the current behaviour.
I will see what can I do for this bug immediately.

**Comment by cmeriaux** [comment]
but it's better to write shortcuts only when they changed. Plus you can edit the shortcuts inside notepad++. For me, my new behaviour is better than the actual behavior (more flexible)

**Comment by sanastasiou** [comment]
If fixing the bug means, that the file is updated on the fly when someone changes a shortcut then by all means, please fix it. Otherwise other plugins, cannot be notified for such changes until Npp is restarted. I actually have a request for exactly this "feature" for one of my plugins we are using at work.

**Comment by cmeriaux** [comment]
@sanastasiou , can you reformul please, I'm not sure to understand what you means ( "then by all means").
Thanks

**Comment by sanastasiou** [comment]
I meant, it would be very useful to have a notification when the user changes a shortcut. If my understanding is correct, currently the shortcut.xml file is only changes upon exiting Notepad++, and plugin developers have no means to actually "ask" about the current configuration other than reading the shortcuts.xml file.

The comment was actually going to @donho .

**Comment by cmeriaux** [comment]
@sanastasiou is your plugin about shortcut management ? If yes, you may develop it directly insinde notepad++
modified shortuct bu user are saved on the array D:\dev\notepad-plus-plus\PowerEditor\src\Parameters.cpp : _customizedShortcuts

**Comment by sanastasiou** [comment]
@cmeriaux Unfortunately no. The plugin adds support for a DSL, called RText. Among others, the language supports comments. Notepad++ has support for shortcuts which add comments to a single line, or more lines. I can reuse the same shortcuts and implement this feature for my plugin, but if the user changes the shortcuts my plugin won't know, until the editor is restarted, because the shortcut.xml is update only upon exit of the editor.
